### PR TITLE
feat: Interfaces generate with associated_types

### PR DESCRIPTION
**Pull request by jawoznia**

### CODE REVIEW

File: sylvia-derive/src/associated_types.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                    Some(ty)
+                }
+                _ => None,
+            })
+            .collect();
+
+        Self(associated_types)
+    }
+
+    pub fn as_where_predicates(&self) -> Vec<WherePredicate> {
```
Comment (kulikthebird) [inline sylvia-derive/src/associated_types.rs:26]: Nitpick: Consider returning an iterator

File: sylvia-derive/src/associated_types.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                let colon = &associated.colon_token;
+                let bound = &associated.bounds;
+                parse_quote! { #name #colon #bound }
+            })
+            .collect()
+    }
+
+    pub fn as_where_clause(&self) -> Option<WhereClause> {
+        let predicates = self.as_where_predicates();
+        match !predicates.is_empty() {
```
Comment (kulikthebird) [inline sylvia-derive/src/associated_types.rs:40]: Nitpick: match --> if

**Review by kulikthebird — CHANGES_REQUESTED** [review]

File: sylvia-derive/src/message.rs [hunk 3]
```diff
@@ -612,10 +522,35 @@ impl<'a> MsgVariant<'a> {
         }
     }
 
+    pub fn emit_trait_querier_impl(&self, associated_name: &[&Ident]) -> TokenStream {
```
[the remark this answers is not in the document]
Comment (jawoznia) [reply sylvia-derive/src/message.rs:525]: This method emits variant declaration and is strictly related to the message variants. As it operates almost entirely on data related to the variants I would leave it here.

We could think about moving the `MsgVariant` somewhere as it's huge on it's own and it's goal is to emit message related methods while other structs like `EnumMessage` are collectors emitting whole messages.

File: sylvia-derive/src/message.rs [hunk 4]
```diff
@@ -787,6 +722,10 @@ impl<'a> MsgVariant<'a> {
         }
     }
 
+    pub fn is_of_type(&self, msg_type: MsgType) -> bool {
```
[the remark this answers is not in the document]
Comment (jawoznia) [reply sylvia-derive/src/message.rs:725]: It's removed in further PR so I will leave it here

File: sylvia-derive/src/associated_types.rs (hunk 1 above)
Comment (jawoznia) [reply sylvia-derive/src/associated_types.rs:26]: It is used only in a single place that's why I decided to make it more "specialized"

File: sylvia-derive/src/associated_types.rs (hunk 2 above)
Comment (jawoznia) [reply sylvia-derive/src/associated_types.rs:40]: ye, makes sense. I guess I like match statement too much ^^

**Review by kulikthebird — APPROVED** [review]

**Pull request merged**